Source/embeddings/Embedding.ts

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { EventType, EventTypeMap } from '@dolittle/sdk.events';
+import {
+    DeleteReadModelInstance,
+    EventSelector,
+    KeySelector,
+    MissingOnMethodForType,
+    ProjectionId
+} from '@dolittle/sdk.projections';
+import { Constructor } from '@dolittle/types';
+import { EmbeddingCompareCallback } from './EmbeddingCompareCallback';
+import { EmbeddingContext } from './EmbeddingContext';
+import { EmbeddingDeleteCallback } from './EmbeddingDeleteCallback';
+import { EmbeddingId } from './EmbeddingId';
+import { EmbeddingProjectCallback } from './EmbeddingProjectCallback';
+import { EmbeddingProjectContext } from './EmbeddingProjectContext';
+import { IEmbedding } from './IEmbedding';
+
+
+/**
+ * Represents an implementation of {@link IEmbedding<T>}.
+ */
+export class Embedding<T> implements IEmbedding<T> {
+
+    /** @inheritdoc */
+    readonly events: Iterable<EventSelector>;
+
+    /**
+     * Initializes a new instance of {@link Embedding}
+     * @param {EmbeddingId} embeddingId The unique identifier for the embedding.
+     * @param {Constructor<T> | T} readModelTypeOrInstance The read model type or instance produced by the embedding.
+     * @param {EventTypeMap<[EmbeddingProjectCallback<any>, KeySelector]>} events The events with respective callbacks and keyselectors used by the embedding.
+     * @param {EmbeddingCompareCallback} _compareMethod The compare method for the embedding.
+     */
+    constructor(
+        readonly embeddingId: EmbeddingId,
+        readonly readModelTypeOrInstance: Constructor<T> | T,
+        private readonly _eventMap: EventTypeMap<[EmbeddingProjectCallback<any>, KeySelector]>,
+        private readonly _compareMethod: EmbeddingCompareCallback,
+        private readonly _deleteMethod: EmbeddingDeleteCallback) {
+        const eventSelectors: EventSelector[] = [];
+        for (const [eventType, [, keySelector]] of this._eventMap.entries()) {
+            eventSelectors.push(new EventSelector(eventType, keySelector));
+        }
+        this.events = eventSelectors;
+    }
+
+    /** @inheritdoc */
+    async on(readModel: T, event: any, eventType: EventType, context: EmbeddingProjectContext): Promise<T | DeleteReadModelInstance> {
+        if (this._eventMap.has(eventType)) {
+            const [method] = this._eventMap.get(eventType)!;
+            return method(readModel, event, context);
+        } else {
+            throw new MissingOnMethodForType(ProjectionId.from(this.embeddingId.value), eventType);
+        }
+    }
+
+    /** @inheritdoc */
+    compare(receivedState: T, currentState: T, context: EmbeddingContext) {
+        return this._compareMethod(receivedState, currentState, context);
+    }
```

Comment:
Where are the parameters different than `(received, current, context)`? The `EmbeddingClassCompareMethod` is the only different one I found by being `(received, context)`, but that's due to the "current" being `this`, the caller of that method. As for why I made it `received` first was that it would look like the `on` methods for projections, where you always get the "old"/"received" read model first and then the event.

Reply:
That was the only case I think, the one you mentioned there. Maybe it isn't such a problem after all